### PR TITLE
Add cloudant specific bulkReplace function

### PR DESCRIPTION
**Pull request by ssh24**
### Description

Implement a cloudant specific bulk update function. Call it `massUpdate`. The function takes in an array of data and performs bulk update against the model instance where the `_rev` of the datas match.

connect to https://github.com/strongloop/loopback-connector-cloudant/issues/35

### CODE REVIEW

**Comment by jannyHou** [comment]
@ssh24 I think as we discussed, this api expects user to provide the `_rev` right? I assume it would be a very simple wrapper of driver function `db.bulkUpdate()`.

- we don't fetch data and merge `_rev` for them
- what ever user provide in `db.connector.massUpdate` is what we pass in to `db.bulkUpdate()`

File: README.md [hunk 1]
```diff
@@ -561,6 +574,68 @@ module.exports = function(server) {
 };
 ```
 
+# Mass update
+
+Given an array of data to be updated, cloudant supports the idea of performing bulk update on a model instance.
```
Comment (dhmlau) [inline README.md]: `cloudant` --> `Cloudant`

File: lib/cloudant.js [hunk 2]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
-        var res = {};
-        res.count = docs.length;
-        return cb(err, res);
+        var errorArray = _.filter(result, 'error');
+        if (errorArray.length > 0) {
+          err = new Error(g.f(util.format('Unable to update 1 or more ' +
+          'document(s): %s', util.inspect(errorArray, 2))));
+          return cb(err);
+        } else {
+          return cb(err, {count: result.length});
```
Comment (dhmlau) [inline lib/cloudant.js]: do we just want to return the count?

**Comment by kjdelisle** [comment]
@jannyHou makes a good point, we should probably rename the function to `bulkReplace`. It gets rid of the ambiguity, uses a more familiar naming convention and also heavily implies the idea that if you don't provide the **whole** object, you'll lose data.

File: README.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ The `loopback-connector-cloudant` module is the Cloudant connector for the LoopB
 - [Discovery](#discovery)
 - [Query](#query)
 - [View](#view)
+- [Mass update](#mass-update)
```
Comment (kjdelisle) [inline README.md]: Let's call it `Bulk replace` instead.

File: README.md [hunk 4]
```diff
@@ -278,6 +279,18 @@ Let's say we have an instance in the database:
     });
   ```
 
+  - update/updateAll
```
Comment (kjdelisle) [inline README.md]: This should be a different PR; this PR is for `bulkReplace`.

File: README.md [hunk 5]
```diff
@@ -561,6 +574,68 @@ module.exports = function(server) {
 };
 ```
 
+# Mass update
+
+Given an array of data to be updated, cloudant supports the idea of performing bulk update on a model instance.
+
+**Note:** To perform mass update, each data in the array data set needs to have a `_rev` property that matches the corresponding documents `_rev` property in the database.
```
Comment (kjdelisle) [inline README.md]: `mass update` -> `bulk replace`

File: README.md [hunk 6]
```diff
@@ -561,6 +574,68 @@ module.exports = function(server) {
...
+        if (err) throw err;
+        console.log('\nCreated instance: ' + JSON.stringify(result));
+
+        dataToUpdate[0]._rev = result[0]._rev;
+        dataToUpdate[1]._rev = result[4]._rev;
+        dataToUpdate[2]._rev = result[7]._rev;
+
+        // note: it is called `db.connector.massUpdate`
+        // rather than `Employee.massUpdate`
+        db.connector.massUpdate('Employee', dataToUpdate, function(err, result) {
```
Comment (kjdelisle) [inline README.md]: `bulkReplace`

File: lib/cloudant.js [hunk 7]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
+};
+
+/**
+ * Perform a bulk update on a model instance
+ *
+ * @param {String} model The model name
+ * @param {Array} datas List of data to be updated
+ * @callback {Function} cb The callback function
+ */
+Cloudant.prototype.massUpdate = function(model, datas, cb) {
```
Comment (kjdelisle) [inline lib/cloudant.js]: `massUpdate` -> `bulkReplace`

File: lib/cloudant.js [hunk 8]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
+  debug('Cloudant.prototype.massUpdate %j %j',
```
Comment (kjdelisle) [inline lib/cloudant.js]: `bulkReplace`

File: lib/cloudant.js [hunk 9]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
+          datas);
+  var self = this;
+  var mo = self.selectModel(model);
+
+  var toBeUpdatedList = [];
+  var data = {};
+
+  self.all(model, {where: {}}, {raw: true}, function(err, docs) {
```
Comment (kjdelisle) [inline lib/cloudant.js]: Unlike the update/updateAll function, we don't want to call `self.all`, since we're just receiving an array of objects to replace in Cloudant.

**Comment by ssh24** [comment]
> If you don't provide the whole object, you'll lose data.

That is not correct. It will do a partial update on whatever data you send it if the **whole** object is not sent.

PS. I can change the name of the function, that is not a big deal.

File: lib/cloudant.js (hunk 2 above)
Comment (ssh24) [inline lib/cloudant.js]: Yes, for `updateAll` we just return the count.

File: lib/cloudant.js [hunk 10]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
+    if (err) return cb(err);
+    async.each(docs, function(doc, cb) {
+      data = _.find(datas, {_rev: doc._rev});
+      if (data) {
+        delete doc._rev;
+        _.mergeWith(doc, data, function(dest, src) { return src; });
+        toBeUpdatedList.push(doc);
+      }
```
Comment (kjdelisle) [inline lib/cloudant.js]: If we do this, it means that only items that have matching `_rev` values will be sent to the database.

While this sounds good at first, what it means is that _the user won't know which values were updated_ since they may have had some of their items dropped from the list without warning.

File: lib/cloudant.js (hunk 2 above)
Comment (b-admike) [inline lib/cloudant.js]: Yeah that's the convention used across the connectors

File: lib/cloudant.js [hunk 11]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
+      });
+    });
+  });
+};
+
+/**
+ * Perform a bulk update on a model instance
+ *
+ * @param {String} model The model name
+ * @param {Array} datas List of data to be updated
```
Comment (b-admike) [inline lib/cloudant.js]: I suggest you use `data` or `dataList` here.

File: lib/cloudant.js [hunk 12]
```diff
@@ -540,11 +540,59 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
+ * @callback {Function} cb The callback function
+ */
+Cloudant.prototype.massUpdate = function(model, datas, cb) {
+  debug('Cloudant.prototype.massUpdate %j %j',
+          datas);
+  var self = this;
+  var mo = self.selectModel(model);
+
+  var toBeUpdatedList = [];
+  var data = {};
```
Comment (b-admike) [inline lib/cloudant.js]: Something along the lines of `dataWithRev` here would be more descriptive

File: README.md (hunk 5 above)
Comment (kjdelisle) [inline README.md]: Given my other feedback, it's also not the case that `_rev` is required; it's only necessary if you know the record exists and want to replace it.

File: test/update.test.js [hunk 13]
```diff
@@ -157,13 +195,89 @@ describe('updateAll', function() {
...
+      testUtil.hasResult(err, result).should.be.ok();
+      dataToBeUpdated[0]._rev = result[0]._rev;
+      dataToBeUpdated[1]._rev = result[3]._rev;
+      dataToBeUpdated[2]._rev = result[5]._rev;
+      db.connector.massUpdate('Product', dataToBeUpdated,
+      function(err, result) {
+        err = testUtil.refinedError(err, result);
+        if (err) return done(err);
+        testUtil.hasResult(err, result).should.be.ok();
+        should.equal(result.length, dataToBeUpdated.length);
```
Comment (b-admike) [inline test/update.test.js]: Do a find here to check everything is good?

Comment (ssh24) [inline test/update.test.js]: Good catch. Added a `find` call underneath.

File: lib/cloudant.js [hunk 14]
```diff
@@ -551,6 +551,36 @@ Cloudant.prototype.updateAll = function(model, where, data, options, cb) {
...
 
 /**
+ * Perform a bulk update on a model instance
+ *
+ * @param {String} model The model name
+ * @param {Array} dataList List of data to be updated
+ * @callback {Function} cb The callback function
+ */
+Cloudant.prototype.bulkReplace = function(model, dataList, cb) {
+  debug('Cloudant.prototype.massUpdate %j %j', model,
```
Comment (kjdelisle) [inline lib/cloudant.js]: change message to `bulkReplace`
